aten/src/ATen/native/cuda/TensorFactories.cu

```diff
@@ -413,4 +419,179 @@ Tensor triu_indices_cuda(
   return tensor;
 }
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ choice ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+__global__ void generate_samples(
+  int64_t *samples,
+  int64_t k,
+  curandStateMtgp32 *state
+){
+  int thread_id = blockIdx.x * blockDim.x + threadIdx.x;
+  samples[thread_id] = curand(state) % (thread_id + k + 1);
+}
+
+template <typename scalar_t>
+__global__ void generate_keys(
+  scalar_t *keys,
+  scalar_t *weights,
+  curandStateMtgp32 *state
+){
+  int thread_id = blockIdx.x * blockDim.x + threadIdx.x;
+  float u = curand_uniform(state);
+  keys[thread_id] = weights[thread_id] > 0? (scalar_t) __powf(u, (float) 1/weights[thread_id]):-1;
+}
+
+__global__ void generate_reservoir(
+  int64_t *indices,
+  int64_t *samples,
+  int64_t nb_iterations,
+  int64_t k
+){
+  for(int i = 0; i < nb_iterations; i++){
+    int64_t z = samples[i];
+    if (z < k) {
+      thrust::swap(indices[z], indices[i + k]);
+    }
+  }
+}
+
+Tensor reservoir_sampling_cuda(
+  const Tensor& x,
+  const Tensor& weights,
+  int64_t k
+){
+
+  AT_CHECK(
+    weights.dtype() == kFloat,
+    "The sampling weights must be Float, got", weights.dtype()
+  );
+
+  AT_CHECK(
+    weights.is_contiguous(),
+    "The sampling weights must be contiguous."
+  );
+
+  int n = x.size(0);
+
+  AT_CHECK(
+    n >= k,
+    "Cannot take a larger sample than population when 'replace=False'"
+  );
+
+  auto options = x.options().dtype(at::kLong);
+  dim3 threads(threadsPerBlock);
+
+  THCState *state = at::globalContext().getTHCState();
+  curandStateMtgp32 *gen_states = THCRandom_generatorStates(state);
+
+  if (weights.numel() == 0){
+    Tensor indices_n = at::arange({n}, options);
+
+    int split, begin, end;
+
+    if(2 * k < n){
+      split = n - k;
+      begin = n - k;
+      end = n;
+    } else {
+      split = k;
+      begin = 0;
+      end = k;
+    }
+
+    int nb_iterations = std::min(k, n - k);
+    dim3 blocks((nb_iterations + threadsPerBlock - 1)/threadsPerBlock);
+
+    Tensor samples = at::arange({nb_iterations}, options);
+
+    generate_samples<<<blocks, threads>>>(
+      samples.data<int64_t>(),
+      split,
+      gen_states
+    );
+
+    generate_reservoir<<<1, 1>>>(
+      indices_n.data<int64_t>(),
+      samples.data<int64_t>(),
+      nb_iterations,
+      split
+    );
+
```

Comment:
Can you add a `AT_CUDA_CHECK(cudaGetLastError());` here and after you called the kernels in the other branch?



AT_CUDA_CHECK(cudaGetLastError());
707

` here and in the other branch?

(note: there is probably a different way of doing this with ATen now, I vaguely remember that something has changed :-)